src/comment_parser.rs

```diff
@@ -0,0 +1,379 @@
+    EndSubscript(usize),
+    /// The start of an italic section `_italic text_`. The `usize` points at the `_` character.
+    StartItalic(usize),
+    /// The end of an italic section `_italic text_`. The `usize` points at the `_` character.
+    EndItalic(usize),
+    /// A bibliographic label `[foo]`. No escapes are needed inside the tag body.
+    BibTag(Span),
+}
+
+impl CommentItem {
+    /// Remove text escapes from a markup segment `buf`, generally coming from the
+    /// [`CommentItem::Text`], [`CommentItem::Label`], or [`CommentItem::Url`] fields.
+    pub fn unescape_text(buf: &[u8], out: &mut Vec<u8>) {
+        let mut iter = buf.iter();
+        while let Some(&c) = iter.next() {
+            match c {
+                b'`' | b'[' | b'~' => {
+                    iter.next();
+                }
+                _ => out.push(c),
+            }
+        }
+    }
+
+    /// Remove math escapes from a markup segment `buf`, generally coming from the
+    /// [`CommentItem::MathToken`] field.
+    pub fn unescape_math(buf: &[u8], out: &mut Vec<u8>) {
+        let mut iter = buf.iter();
+        while let Some(&c) = iter.next() {
+            match c {
+                b'`' => {
+                    iter.next();
+                }
+                _ => out.push(c),
+            }
+        }
+    }
+
+    const fn token(math: bool, span: Span) -> Self {
+        if math {
+            Self::MathToken(span)
+        } else {
+            Self::Text(span)
+        }
+    }
+}
+
+/// An iterator over a metamath text comment, yielding markup items.
+#[derive(Debug, Clone)]
+pub struct CommentParser<'a> {
+    buf: &'a [u8],
+    pos: usize,
+    math_mode: bool,
+    html_mode: bool,
+    item: Option<CommentItem>,
+    end_italic: usize,
+    end_subscript: usize,
+}
+
+impl<'a> CommentParser<'a> {
+    /// Construct a new `CommentParser` from a sub-span of a buffer.
+    /// The returned comment items will have spans based on the input span,
+    /// and the portion of the buffer outside `buf[span.start..span.end]` will be ignored.
+    #[must_use]
+    pub fn new(buf: &'a [u8], span: Span) -> Self {
+        Self {
+            buf: &buf[..span.end as usize],
+            pos: span.start as _,
+            math_mode: false,
+            html_mode: false,
+            item: None,
+            end_italic: usize::MAX,
+            end_subscript: 0,
+        }
+    }
+
+    /// Remove text escapes from a markup segment `span`, generally coming from the
+    /// [`CommentItem::Text`], [`CommentItem::Label`], or [`CommentItem::Url`] fields.
+    pub fn unescape_text(&self, span: Span, out: &mut Vec<u8>) {
+        CommentItem::unescape_text(span.as_ref(self.buf), out)
+    }
+
+    /// Remove math escapes from a markup segment `span`, generally coming from the
+    /// [`CommentItem::MathToken`] field.
+    pub fn unescape_math(&self, span: Span, out: &mut Vec<u8>) {
+        CommentItem::unescape_math(span.as_ref(self.buf), out)
+    }
```

Comment:
Since `unescape_text` and `unescape_math` have to be called on the `CommentItem`, why not provide the API there?
A single function could provide the unescaped result depending on the context (math unescaped for math, "text" unescaped for text, label, url.

Reply:
There are functions on both types, but the one on `CommentItem` has no context to interpret the spans with.

Reply:
I considered having wrappers like `struct TextSpan(pub Span)` to make it easier to get the unescaping right, but I don't want to tie the client's hands too much on how they want to handle this data.